### PR TITLE
PDA UI refactor and cartridges

### PR DESCRIPTION
**Pull request by juliangiebel**
## About the PR 
This PR significantly improves the PDA UI, adds colored borders to the UI that match the PDAs colors and adds cartridges.

Cartridges are entities you can stick into a PDA and the PDA can "run" the program that is on the cartridge. A cartridge entity provides a UI fragment to the `CartridgeLoaderSystem` (the PDA) which gets displayed inside the PDAs UI when the program is active.

Programs are just systems listening to events for their component on a cartridge entity.
Programs can send PDA notifications (This will most likely be implemented in a future PR to keep the scope down)

Cartridges can be installed which spawns their prototype into a container on the PDA. Programs can run in the Background which means they will receive relayed events and can send PDA notifications even though they are not the active program.

The `CartridgeLoaderSystem` relays some events to the cartridge entity: 
- `AfterInteractEvent`,
- `NetworkPacketReceivedEvent`
- `BoundUserInterfaceMessage`

`BoundUserInterfaceMessage` doesn't get relayed to background programs or when the message is actually meant for the `CartridgeLoaderSystem`.

This PR also adds a single cartridge as a reference and for testing the `CartridgeLoaderSystem`, the notekeeper cartridge.

There are still some things that need to be done but the `CartridgeLoaderSystem` itself can be reviewed already.

- [x] Finish implementing the notekeeper cartridge
- [x] Preinstalled programs
- [ ] Set the remaining PDA border colors
- [x] Doc comments!
- [ ] Finding good cartridge sprites
- [ ] Documentation for creating cartridges

Finding and setting the remaining PDA colors and finding good cartridge sprites are both thing I could use some help with.

**Screenshots**

### Some PDA colors and the home screen
image
### Empty programs list. Installed programs and the inserted cartridge would show up here.
image
### The settings screen (Crew manifest, instrument, ringtone and the uplink moved here)
image
### The unfinished notekeeper UI with the titlebar that gets displayed which is part of the PDA UI
image
### The programs list with a cartridge inserted
image

**Changelog**

:cl:
- add: a fancy new PDA UI
- add: cartridges containing Programs for your PDA

### CODE REVIEW

File: Resources/Prototypes/Entities/Objects/Devices/pda.yml [hunk 1]
```diff
@@ -41,6 +42,22 @@
...
+      priority: -1
+      name: Cartridge
+      ejectSound: /Audio/Machines/id_swipe.ogg
+      insertSound: /Audio/Weapons/Guns/MagIn/batrifle_magin.ogg
+      whitelist:
+        components:
+          - Cartridge
   - type: ActivatableUI
     key: enum.PDAUiKey.Key
     singleUser: true
```
Comment (theashtronaut) [inline Resources/Prototypes/Entities/Objects/Devices/pda.yml:63]:
```suggestion
    singleUser: true
    closeOnHandDeselect: false
```
Might be a nice time to add this to the ActivationUI now that it's a possibility. This makes it so you can switch the active hand and the PDA UI would remain up, which would likely be useful for keeping an eye on whatever you are doing on your PDA while multitasking

**Comment by juliangiebel** [comment]
This is ready for review now.
The few missing border colors can be added later and cartridge sprites can be added when a proper cartridge gets added.

I tried my best to split this PR up into just the UI stuff for PDAs and cartridges after the fact but I failed because the PDA UI depends heavily on the cartridge loader UI classes, so the PR got a bit large. 😔 

I decided against leaving out the Notekeeper cartridge as it serves as an example implementation and without it the cartridge loader couldn't be tested at all.

File: Content.Server/CartridgeLoader/Cartridges/NotekeeperCartridgeSystem.cs [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+    /// <summary>
+    /// The ui messages received here get wrapped by a CartridgeMessageEvent and are relayed from the <see cref="CartridgeLoaderSystem"/>
+    /// </summary>
+    /// <remarks>
+    /// The cartridge specific ui message event needs to inherit from the CartridgeMessageEvent
+    /// </remarks>
+    private void OnUiMessage(EntityUid uid, NotekeeperCartridgeComponent component, CartridgeMessageEvent args)
+    {
+        if (args is not NotekeeperUiMessageEvent message)
```
Comment (Macoron) [inline Content.Server/CartridgeLoader/Cartridges/NotekeeperCartridgeSystem.cs:33]: I'm also not sure about this relayed UI events from bound user interface. If you used components on a PDA entity, you could subscribe directly to them without any relay.

**Review by Macoron — COMMENTED** [review]
I tested in game, works good. This can be reused for computers in future. Good job!

File: Content.Server/CartridgeLoader/CartridgeLoaderSystem.cs [hunk 3]
```diff
@@ -0,0 +1,413 @@
...
+            return false;
+
+        if (_containerSystem?.TryGetContainer(loaderUid, ContainerName, out var container) != true)
+            return false;
+
+        //Prevent installing cartridges that have already been installed
+        if (IsInstalled(prototype, loader))
+            return false;
+
+        var installedProgram = Spawn(prototype, new EntityCoordinates(loaderUid, 0, 0));
```
[the remark this answers is not in the document]
Comment (juliangiebel) [reply Content.Server/CartridgeLoader/CartridgeLoaderSystem.cs:134]: There are multiple reasons.
I can use entity uids for the programs for example. Managing what event goes to which program is also easier. And moony told me to make them entities.
I habe the conversation as a screenshot in this document: https://hackmd.io/DwBGv7e7T5qkv9uWK5jcaQ

**Comment by Cheackraze** [comment]
there are unused cartridge sprites in the files \Resources\Textures\Objects\Devices\cartridge.rsi but they arent amazing

**Comment by juliangiebel** [comment]
I made a doc explaining some of the concepts behind cartridges:
https://docs.spacestation14.io/en/content/cartridge-loader

File: Content.Server/CartridgeLoader/CartridgeLoaderSystem.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (juliangiebel) [reply Content.Server/CartridgeLoader/CartridgeLoaderSystem.cs:134]: Can I mark these questions as resolved?

**Pull request merged**